### PR TITLE
C++, add support for parameterized noexcept specifier in function dec…

### PR DESCRIPTION
**Pull request by jbab**
…larations
Subject: C++, Add support for parameterized noexcept specifier in function declarations

### Feature or Bugfix
- Feature

### Purpose

A function declaration such as

    cpp:function:: vector(vector&&) noexcept(std::is_nothrow_move_constructible_v<T>)

lead to a parse failure with the message _Parameterised "noexcept" not yet implemented._

This PR implements the missing feature,

### CODE REVIEW

**Comment by jakobandersen** [comment]
Thanks a lot for your effort. However, I do not believe that ``ASTNoexceptExpr`` is the right element to adapt here as it is for the other kind of noexcept. Instead I suggest to add as new ``ASTNoexceptSpec`` (located nearby the AST class for parameters and qualifiers), which models ``noexcept`` or ``noexcept(<constant-expression>)``.

File: sphinx/domains/cpp.py [hunk 1]
```diff
@@ -5495,11 +5514,14 @@ def _parse_parameters_and_qualifiers(self, paramMode: str) -> ASTParametersQuali
         initializer = None
         self.skip_ws()
         if self.skip_string('noexcept'):
-            exceptionSpec = 'noexcept'
-            self.skip_ws()
-            if self.skip_string('('):
-                self.fail('Parameterised "noexcept" not yet implemented.')
-
+            if self.skip_string_and_ws('('):
+                expr = self._parse_expression()
```
Comment (jakobandersen) [inline sphinx/domains/cpp.py]: Please change this to use ``_parse_constant_expression()``.

File: tox.ini [hunk 2]
```diff
@@ -17,7 +17,7 @@ description =
     py{35,36,37,38,39}: Run unit tests against {envname}.
     du{12,13,14}: Run unit tests with the given version of docutils.
 deps =
-    git+https://github.com/html5lib/html5lib-python  # refs: https://github.com/html5lib/html5lib-python/issues/419
+    git+https://github.com/html5lib/html5lib-python# refs: https://github.com/html5lib/html5lib-python/issues/419
```
Comment (jakobandersen) [inline tox.ini]: Looks like a stray change.

**Comment by jakobandersen** [comment]
Thanks for the quick work. I made a two comments for some specific code, but otherwise it looks good. As a last request, it would be great if you could squash the commits to a single one in the end to avoid touching too many lines that were anyway changed back.

File: sphinx/domains/cpp.py (hunk 1 above)
Comment (jbab) [reply sphinx/domains/cpp.py]: Was modeled according to `def _parse_unary_expression` (line 5031). Should it be changed there as well?

File: tox.ini (hunk 2 above)
Comment (jbab) [reply tox.ini]: No it was not. In my environment (Ubuntu WSL on Windows 10) tox interpreted the whole string up to the `#` as URL. This change (or deleting the comment completely) was the only way I could run tox.
I admit I did not research further to find alternative solutions or the real root cause.

Comment (jakobandersen) [reply tox.ini]: Ah, ok. Perhaps @tk0miya knows more about the tox stuff?

File: sphinx/domains/cpp.py (hunk 1 above)
Comment (jakobandersen) [reply sphinx/domains/cpp.py]: No, only here. I try to stay close to the grammar in the standard, which here says constant-expression, which excludes comma expressions and assignment expressions.

File: tox.ini (hunk 2 above)
Comment (tk0miya) [reply tox.ini]: I can't understand what happened in @jbab's environment. But, I feel this change is strange. I wrote it as a comment, not a part of URL. -1 for this change.

Comment (jakobandersen) [reply tox.ini]: Right, it's a bit odd. @jbab, if you change this part back and submit a separate issue/PR to discuss this, then we can get the C++ part merged. Thanks.

Comment (jbab) [reply tox.ini]: ok, done.

**Comment by jakobandersen** [comment]
Thanks!

**Pull request merged**